### ISSUE TITLE
08-k-nearest-neighbours-winequality-red.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/proyecto-machine-learning-gregoriant#6]
Hola @gregangel1550-del , por aquí las issues!

## LO QUE HAS HECHO BIEN

- **Imports completos**: Tienes lo necesario para el ejercicio.

- **Análisis exploratorio**: shape, info(), describe(), value_counts().

- **Verificación de duplicados**: Usaste `.duplicated().sum()` para contar.

- **Heatmap de correlaciones**: Buen análisis visual.

- **Train/test split**: Correcto con 80/20.

- **Hiperparametrización**: Usaste RandomizedSearchCV con varios parámetros.

- **Evaluación del modelo optimizado**: Usaste `best_estimator_` correctamente.

---

### **1. El drop_duplicates NO se guardó - ERROR CRÍTICO**
```python
df.drop_duplicates()  # ← Falta inplace=True o reasignar
```

Sin `inplace=True` o `df = df.drop_duplicates()`, los duplicados **siguen en el DataFrame**. Tienes 240 duplicados que no eliminaste.

---

### **2. NO creaste la variable "label" - ERROR CRÍTICO**
El enunciado dice claramente:
> *"La columna objetivo es label: 0 = Baja calidad, 1 = Calidad media, 2 = Alta calidad"*

Tú usaste `quality` directamente (que tiene valores 3-8), en lugar de crear `label` con 3 clases:

```python
df["label"] = df["quality"].apply(lambda q: 0 if q <= 4 else (1 if q <= 6 else 2))
```

Esto es importante porque el ejercicio pide clasificar en **3 categorías**, no en las 6 originales.

---

### **3. Escalaste el DataFrame equivocado**
```python
norm_features = norm_scaler.fit_transform(df[num_variables])  # ← Escala df original
```

Pero luego entrenas con:
```python
model.fit(X_train, y_train)  # ← X_train NO está escalado
```

**El modelo entrena con datos SIN escalar**. KNN es muy sensible a la escala de las variables. Deberías:
```python
X_train_scaled = norm_scaler.fit_transform(X_train)
X_test_scaled = norm_scaler.transform(X_test)
model.fit(X_train_scaled, y_train)
```

---

### **4. NO hiciste la gráfica de accuracy vs k**
El enunciado pedía:
> *"Crea un bucle para probar diferentes valores de k. Grafica accuracy vs k"*

No hay ninguna gráfica en tu notebook.

---

### **5. NO usaste confusion_matrix ni classification_report**
El enunciado pedía evaluar con estas métricas además de accuracy.

---

### **6. NO creaste la función predict_wine_quality**
El enunciado pedía como bonus crear una función de predicción.

---

### **7. NO guardaste el modelo**
No hay código con pickle para guardar el modelo final.

---

### **8. Creaste df_optimized pero no lo usaste**
```python
df_optimized = df.drop(columns=['residual sugar', 'free sulfur dioxide', 'pH', 'fixed acidity'], axis=1)
```

Creaste este DataFrame eliminando columnas, pero después usas `df` para el split. ¿Para qué lo creaste?

---

**Greg**, el ejercicio está bastante incompleto. Tienes varios errores críticos: no aplicaste el drop de duplicados, no creaste la variable `label` que pedía el enunciado, y **entrenaste el modelo con datos sin escalar** (esto es grave para KNN). Además te faltan la mayoría de los requisitos del enunciado.

**Comment by mezcolantriz** [comment]
@gregangel1550-del 

## CORRECCIONES APLICADAS

| Error Original | Estado |
|----------------|--------|
| drop_duplicates no se guardaba | ✅ Ahora `df = df.drop_duplicates()` |
| No creó la variable label (3 clases) | ✅ Ahora usa `np.select()` para crear 0, 1, 2 |
| Creó df_optimized pero no lo usaba | ✅ Ahora sí lo usa para el split |
| No usó confusion_matrix | ✅ Ahora tiene `ConfusionMatrixDisplay` |
| No guardó el modelo | ✅ Ahora guarda con `dump()` |

---

## ERRORES QUE AÚN PERSISTEN

### 1. El escalado está MAL en RandomizedSearchCV y predicciones finales - ERROR GRAVE

El modelo base lo entrenas bien con datos escalados:

```python
X_train_scaled = norm_scaler.fit_transform(X_train)
X_test_scaled = norm_scaler.transform(X_test)
model.fit(X_train_scaled, y_train)  #  Correcto
```

**PERO** luego en RandomizedSearchCV:
```python
random_search.fit(X_train, y_train)  # 👎  X_train SIN escalar!
```

Y las predicciones finales:
```python
y_pred = best_knn.predict(X_test)  # 👎  X_test SIN escalar!
y_pred_test = random_search.predict(X_test)  # 👎  X_test SIN escalar!
```

**Esto invalida TODA la hiperparametrización.** Debe ser:
```python
random_search.fit(X_train_scaled, y_train)
# ...
y_pred = best_knn.predict(X_test_scaled)
```

---

### 2. Código de escalado duplicado e innecesario

Tienes esta celda que NO sirve para nada:
```python
norm_scaler = StandardScaler()
num_variables = ['volatile acidity','citric acid',...]
norm_features = norm_scaler.fit_transform(df_optimized[num_variables])
df_norm = pd.DataFrame(norm_features, ...)
```

`df_norm` no se usa nunca. Y  haces `fit_transform` sobre TODO el dataset (data leakage) y luego vuelve a hacer `fit_transform` sobre X_train. El scaler se sobreescribe pero es confuso.

---

### 3. Falta la gráfica de accuracy vs k
El enunciado pedía:
> "Crea un bucle para probar diferentes valores de k. Grafica accuracy vs k"

**NO está.**

---

### 5. Falta función predict_wine_quality
No necesario pero si recomendable

---

## RESUMEN FINAL

| Aspecto | Estado |
|---------|--------|
| drop_duplicates aplicado | ✅ Corregido |
| Variable label (3 clases) | ✅ Corregido |
| Usar df_optimized | ✅ Corregido |
| Modelo base con datos escalados | ✅ Corregido |
| RandomizedSearchCV con datos escalados | ❌ USAS DATOS SIN ESCALAR |
| Predicciones con datos escalados | ❌ USAS DATOS SIN ESCALAR |
| confusion_matrix | ✅ Añadido |
| classification_report | ❌ Falta |
| Gráfica accuracy vs k | ❌ Falta |
| Guardar modelo | ✅ Corregido |
| Guardar scaler | ❌ Falta |
| Función predict | ❌ Falta |

---

Has corregido varias cosas, pero el **error más grave persiste**: la hiperparametrización y predicciones finales usan datos SIN escalar. Todo el trabajo de RandomizedSearchCV está "contaminado". Solo el modelo base (sin optimizar) está bien entrenado.

El notebook necesita otra revisión para que use `X_train_scaled` y `X_test_scaled` en TODAS las operaciones después del escalado.
